src/EditorFeatures/Core/Shared/Preview/PreviewSolutionCrawlerRegistrationService.cs

```diff
@@ -1,72 +1,112 @@
 
-                var analyzer = _provider.CreateIncrementalAnalyzer(workspace);
-                var source = s_cancellationTokens.GetValue(workspace, _ => new CancellationTokenSource());
+                var solution = _workspace.CurrentSolution;
+                var documentIds = _workspace.GetOpenDocumentIds().ToImmutableArray();
 
-                var solution = workspace.CurrentSolution;
-                foreach (var documentId in workspace.GetOpenDocumentIds())
+                try
                 {
-                    var document = solution.GetDocument(documentId);
-                    if (document == null)
+                    foreach (var documentId in documentIds)
                     {
-                        continue;
-                    }
+                        var document = solution.GetDocument(documentId);
+                        if (document == null)
+                        {
+                            continue;
+                        }
 
-                    // delay analyzing
-                    await Task.Delay(workerBackOffTimeSpanInMS).ConfigureAwait(false);
+                        // delay analyzing
+                        await Task.Delay(workerBackOffTimeSpanInMS, _source.Token).ConfigureAwait(false);
 
-                    // do actual analysis
-                    await analyzer.AnalyzeSyntaxAsync(document, source.Token).ConfigureAwait(false);
-                    await analyzer.AnalyzeDocumentAsync(document, bodyOpt: null, cancellationToken: source.Token).ConfigureAwait(false);
+                        // do actual analysis
+                        await diagnosticAnalyzer.AnalyzeSyntaxAsync(document, _source.Token).ConfigureAwait(false);
+                        await diagnosticAnalyzer.AnalyzeDocumentAsync(document, bodyOpt: null, cancellationToken: _source.Token).ConfigureAwait(false);
 
-                    // don't call project one.
+                        // don't call project one.
+                    }
+                }
+                catch (OperationCanceledException)
+                {
+                    // do nothing
                 }
             }
-            catch (OperationCanceledException)
-            {
-                // do nothing
-            }
-        }
 
-        public void Unregister(Workspace workspace, bool blockingShutdown = false)
-        {
-            CancellationTokenSource source;
-            if (s_cancellationTokens.TryGetValue(workspace, out source))
+            public async void Unregister(Workspace workspace, bool blockingShutdown = false)
             {
-                source.Cancel();
+                Contract.ThrowIfFalse(workspace == _workspace);
+                _source.Cancel();
+
+                // wait for analyzer work to be finished
+                await _analyzeTask.ConfigureAwait(false);
+
+                // ask it to reset its stages for the given workspace
+                _owner._analyzerService.ShutdownAnalyzerFrom(_workspace);
```

Comment:
My point wasn't that i cared about the name.  it was that i was uncertain if this method was on the right object.  i.e. you're sayign "AnalyzerService remove the cached data for this workspace".  Rather than just asking the analyzer itself to do that.  Since you created the analyzer, it made sense to me that you would then be the one asking it to clean up.

Now: i could def be wrong here, and putting this functionality on the service may be the right thing.   However, i'm looking for information/explanation to help me understand why this is the right location. 

Thanks!